Fix the behavior of 'last-focused-main-window' and use it for new (IPC) tabs

haasn opened the pull request:
Fixes a bunch of “silent” bugs, and one feature request.

I would like to add tests to this but I'm not entirely sure how. (I would need to switch focus, then spawn a new tab via the IPC, then see if the right window got the tab)

haasn commented:
Re: tests: never mind, I found some tests that I teach me how to implement this. (I think)

haasn commented:
Turns out that this change, when running inside `xvfb`, breaks some tests which, as written, explicitly depend on the fact that new tabs in new windows are opened in the “last created” window rather than the last focused window by default.

I'll have to rewrite those tests, since it's a bug in the test and not in the functionality (it works fine, but the implicit assumption that `:open -w` also focuses the new window is false inside `xvfb`)

haasn commented:
Just noting this down for the record:

The tests broken inside xvfb by this change include:

* features/search.feature:99
* features/search.feature:140
* features/tabs.feature:868
* features/tabs.feature:831

In the first two cases, the offending command is `Open X in a new window`, which does not focus that window. In the second two cases, the offending command is `:open -w URL`.

Not actually sure how to fix these, since I don't know how to tell X to force-focus a window.

The-Compiler commented on qutebrowser/app.py:
```diff
@@ -343,15 +343,15 @@ def on_focus_changed(_old, new):
-    else:
+    # Find the newly focused widget's parent MainWindow, if it exists
+    while new is not None and not isinstance(new, mainwindow.MainWindow):
+        new = new.parent()
```
Couldn't you just get the `QApplication.activeWindow()` here?

haasn replied:
Of course, good point. I wrote that bit before discovering activeWindow(). :p

haasn commented:
Updated to fix the issue with tests (solved by introducing a new setting which we can then force-set)

The-Compiler commented on qutebrowser/app.py:
```diff
@@ -343,15 +343,12 @@ def on_focus_changed(_old, new):
-        qApp.restoreOverrideCursor()
-    else:
-        objreg.register('last-focused-main-window', new.window(), update=True)
+    if new is not None:
```
You removed the `isinstance` check here, which means the mouse cursor would remain hidden when switching focus to a dialog box or something similar.

I'd propose the following:

```python
if not isinstance(new, mainwindow.MainWindow):   # includes an implicit None check anyways
    qApp.restoreOverrideCursor()
window = qApp.activeWindow()
if window is not None:
    objreg.register(...)
```

The-Compiler replied:
Hmm, though I see another issue here: `qapp.activeWindow()` could actually be the web inspector, or a dialog box, or another top-level window... And apparently we can't get the toplevel `QWidget` from a `QWindow`?

So I guess the real solution would be to get [`new.window()`](http://doc.qt.io/qt-5/qwidget.html#window) (which is actually the toplevel `QWidget`, fun...) and register that if it's a `mainwindow.MainWindow` instance.

haasn replied:
> You removed the isinstance check here, which means the mouse cursor would remain hidden when switching focus to a dialog box or something similar.

Re-including this check would break `_maybe_hide_mouse_cursor()` again - because the issue is that a `MainWindow` never gets directly focused (at least on my end), it's always some child - in this case the webkit-specific WebView.

So unless we have some good way of distinguishing between if what I clicked was an object that should cause the cursor to be re-shown or not, I don't see a way to reimplement this feature in a way that actually works. (Since the current almost surely doesn't)

haasn commented:
Updated to remove the buggy `hide-mouse-cursors` option as I suggested, and also squashed/changed some commits to both clean up the PR and address the mentioned issues / build failures.

haasn commented:
> I'm not sure if I like the idea of removing it, but I guess you're right...

You could also omit that commit, keep the new behavior where `auto-hide-cursor` means “always hide the cursor”, and see if somebody complains about that not un-hiding it for input fields.

haasn commented:
> Looking at f77ba57 there are also some more bits in modeman.py to clean up.

That code seems to no longer be present at all.

> You'll also need to add it to REMOVED_OPTIONS in config.py so it gets removed from existing configs instead of qutebrowser complaining about it.

Updated.

The-Compiler commented on qutebrowser/config/configdata.py:
```diff
@@ -227,6 +227,16 @@ def data(readonly=False):
+            ('new-instance-open-target.window',
+             SettingValue(typ.String(
+                 valid_values=typ.ValidValues(
+                     ('last-opened', "Open new tabs in the last"
```
You're missing a space before the `"` here and below, but I'll fix it up when merging.

The-Compiler commented:
Thanks!
